### ISSUE TITLE
Support assignment of stereochemistry tags to bonds from 3D structure

### ISSUE THREAD

**Issue by gncs** [rdkit/rdkit#1497]
RDKit 2017.3.03, GCC 6.2.0

To assign stereochemistry to a molecule with a 3D conformer I run:
```c++
RDKit::MolOps::findPotentialStereoBonds(mol, true);
RDKit::MolOps::assignChiralTypesFrom3D(mol);
RDKit::MolOps::assignStereochemistry(mol, true, true);
```

This does work for the chiral type of atoms (R/S) but I cannot get it to work for bonds (E/Z or cis/trans).
All I am getting is `RDKit::Bond::STEREOANY` for double bonds.
Note: I built the molecule "from scratch", i.e., I added atoms and bonds manually.

What am I missing here?

**Comment by pemsley** [comment]
Is Chirality::findAtomNeighborDirHelper() being called with hasExplicitUnknownStereo set true?

**Comment by gncs** [comment]
@pemsley, could you elaborate on that? which function am I missing? the functions listed above are the only ones I am using that concern chirality or stereo information.

**Comment by greglandrum** [comment]
I agree that there should be a way to do this.
It seems like we need a new function: `RDKit::MolOps::assignStereochemistryFrom3D()` that handles both atomic and double bond stereochemistry.

**Comment by gncs** [comment]
@greglandrum, thank you for adding this feature!

Could you give me an ETA for this feature? Or maybe describe to me how I would implement that myself? I am asking because I need this functionality rather urgently :)
Thanks!

**Comment by pemsley** [comment]
@gncs, I, in the past, have been constructing molecules "from scratch" myself. You stirred some memory that atoms (maybe  bonds) need to have their stereochemistry set to "unassigned/unset" before assignStereochemistry() would do what you/I wanted. The critical path, it seems to me (inside of assignStereochemistry()) depended on the state of hasExplicitUnknownStereo - I was encouraging you to poke inside RDKit's innards :-).

**Comment by pemsley** [comment]
@gncs: for the double bond dblBond between B=C, 
find the non-Hydrogen atoms A-B=C-D,
Use RDGeom::Point3D &r_pos = conf.getAtomPos(iat) to get the coords for A,B,C,D. 
dp = DotProduct((B-A), (C-D));
if (dp > 0) 
   dblBond->setStereo(Bond::STEREOZ);
else
   dblBond->setStereo(Bond::STEREOE);

**Comment by greglandrum** [comment]
The feature would be part of the September release, but it'll show up here on Github as soon as it's done.
In the meantime, you can add a call to `DetectBondStereochemistry()`, which is declared in `Code/GraphMol/FileParsers/MolFileStereochem.cpp` to your code. This can come either before or after the call to `RDKit::MolOps::assignChiralTypesFrom3D(mol);`

**Comment by gncs** [comment]
@greglandrum, thank you for this hint!

I now have the following:
```c++
RDKit::DetectBondStereoChemistry(mol, &mol.getConformer());
RDKit::MolOps::assignChiralTypesFrom3D(mol);
RDKit::MolOps::assignStereochemistry(mol, true, true);
```
Note, that I removed the call to `RDKit::MolOps::findPotentialStereoBonds()` as this sets the stereo of the double bond to `Bond::STEREOANY` and `DetectBondStereoChemistry()` skips those.

I am running a couple of test, but it seems as if this fixes my problem -- looking forward to the new release!

**Comment by greglandrum** [comment]
Closed in #1555
